### PR TITLE
Modifications to load_novel_text() structure

### PR DESCRIPTION
**Pull request by ArrowAce**
I separated the load_novel_text() function into distinct parts for ease of understanding, and also out of the recognition that there are more ways to clean the text than just removing the boilerplate. Separating the different cleaning methods into distinct functions makes it easier to add new functions to clean the text and edit existing functions.

I moved the boilerplate remover into a separate function and also changed the way the Boilerplate remover works as some books did not follow standard Gutenberg format; the new function works by using an existing boilerplate remover in the gutenberg module.

I created a Table of Contents removing function that I am working at the moment; it is marked with a TODO.

### CODE REVIEW

File: gender_novels/corpus_gen.py [hunk 1]
```diff
@@ -57,8 +57,16 @@ def generate_corpus_gutenberg():
         author = get_author_gutenberg(gutenberg_id)
         novel_metadata['author'] = author
         title = get_title_gutenberg(gutenberg_id)
+<<<<<<< HEAD
```
Comment (srisi) [inline gender_novels/corpus_gen.py]: There's a merge conflict here, indicated by `<<<<<<<< HEAD`, meaning that this file contains code from both your local repo and the upstream master. Please make sure to resolve it before committing again. (the easiest solution might be to just copy corpus_gen.py from the upstream master.

File: gender_novels/novel.py [hunk 2]
```diff
@@ -120,12 +121,13 @@ def __str__(self):
...
+    def load_novel_text(self):
+        """Loads the text of a novel and uses the remove_boilerplate_text() and
+        remove_table_of_contents() functions on the text of the novel to remove the boilerplate
+        text and table of contents from the novel. After these actions, the novel's text should be
+        only the actual text of the novel.
 
-        Currently only supports boilerplate removal for Project Gutenberg ebooks.
-
-        :rtype: str
+        :return: string
```
Comment (srisi) [inline gender_novels/novel.py]: the actual return type should be `str` (`string` is not a Python type, `str` is.)

File: gender_novels/novel.py [hunk 3]
```diff
@@ -120,12 +121,13 @@ def __str__(self):
         return name
 
 
-    def _load_novel_text(self):
-        """Loads the text of a novel and removes boilerplate at the beginning and end
+    def load_novel_text(self):
```
Comment (srisi) [inline gender_novels/novel.py]: The function should still be private (starting with an underscore). The way to access the text should be via the Novel.text attribute.

File: gender_novels/novel.py [hunk 4]
```diff
@@ -120,12 +121,13 @@ def __str__(self):
...
+        """Loads the text of a novel and uses the remove_boilerplate_text() and
+        remove_table_of_contents() functions on the text of the novel to remove the boilerplate
+        text and table of contents from the novel. After these actions, the novel's text should be
+        only the actual text of the novel.
```
Comment (srisi) [inline gender_novels/novel.py]: This function needs a doctest.

File: gender_novels/novel.py [hunk 5]
```diff
@@ -137,17 +139,72 @@ def _load_novel_text(self):
...
+
+        return text
+
+
+    def remove_boilerplate_text(text):
+        """
+        Removes the boilerplate text from an input string of a novel.
+        Currently only supports boilerplate removal for Project Gutenberg ebooks. Uses the
+        strip_headers() function from the gutenberg module.
+
```
Comment (srisi) [inline gender_novels/novel.py]: Needs doctest.

File: gender_novels/novel.py [hunk 6]
```diff
@@ -137,17 +139,72 @@ def _load_novel_text(self):
...
-        # Extract Project Gutenberg Boilerplate
-        if text.find('*** START OF THIS PROJECT GUTENBERG EBOOK') > -1:
-            end_intro_boilerplate = text.find(
-                '*** START OF THIS PROJECT GUTENBERG EBOOK')
-            # second set of *** indicates start
-            start_novel = text.find('***', end_intro_boilerplate + 5) + 3
-            end_novel = text.find('*** END OF THIS PROJECT GUTENBERG EBOOK')
-            text = text[start_novel:end_novel]
+        # These two functions will ensure that the text becomes only the novel's own text,
+        # removing the Project Gutenberg .
```
Comment (srisi) [inline gender_novels/novel.py]: ... boiler plate?

File: gender_novels/novel.py [hunk 7]
```diff
@@ -137,17 +139,72 @@ def _load_novel_text(self):
...
+        text = self.remove_boilerplate_text(text)
+        text = self.remove_table_of_contents(text)
+
+        return text
+
+
+    def remove_boilerplate_text(text):
```
Comment (srisi) [inline gender_novels/novel.py]: This function should include the existing Project Gutenberg boiler plate removal:
```
        if text.find('*** START OF THIS PROJECT GUTENBERG EBOOK') > -1:
            end_intro_boilerplate = text.find(
                '*** START OF THIS PROJECT GUTENBERG EBOOK')
            # second set of *** indicates start
            start_novel = text.find('***', end_intro_boilerplate + 5) + 3
            end_novel = text.find('*** END OF THIS PROJECT GUTENBERG EBOOK')
            text = text[start_novel:end_novel]
```

File: gender_novels/novel.py [hunk 8]
```diff
@@ -137,17 +139,72 @@ def _load_novel_text(self):
...
+        followed only by empty lines until finding a line which is not entirely allcaps.
+
+        :return: string
+        TODO(derek): Make this fully functional
+        """
+
+        # Finds all positions of line breaks in the text file for the first few lines, tuned off
+        # the parameter of charactersearch characters out
+        charactersearch = 8000
+        line_breaks = text[0:charactersearch].find("\n")
```
Comment (srisi) [inline gender_novels/novel.py]: This only returns the position of the first line break.

File: gender_novels/novel.py [hunk 9]
```diff
@@ -137,17 +139,72 @@ def _load_novel_text(self):
...
+        # Creates a 2D array where each entry is the [contents of line, position of the line's
+        # start, position of the line's end]. The last line break is not considered because,
+        # of course, there is no next line after the last line.
+        lines_of_text = []
```
Comment (srisi) [inline gender_novels/novel.py]: One possible way of splitting up the text into individual lines would be to use `text.split('\n')` That way, you wouldn't have to tweak the line_breaks search above.

**Comment by mscuthbert** [comment]
Good ideas all around, but it looks like there's a merge conflict that needs fixing in corpus_gen.py

**Comment by mscuthbert** [comment]
the private vs public difference of opinion on _load_novel_text() just shows that people who have programmed for a long time can still disagree on the distinction between public and private.  Follow Stephan on this since he made the initial object and accept apologies from me.

File: gender_novels/novel.py (hunk 8 above)
Comment (mscuthbert) [reply gender_novels/novel.py]: re.findall?

File: gender_novels/novel.py (hunk 9 above)
Comment (mscuthbert) [reply gender_novels/novel.py]: but look at a recent pull-request that wanted to remove all of them (the NLTK request?) and discuss with that author.

**Comment by srisi** [comment]
Last follow-up on public vs private: If there is a reason why you'd want to access remove_boiler_plate() or load_novel_text() outside of the Novel object itself, e.g. in corpus_gen.py, then that function should be public. But as you can see from our internal discussion on this topic, this is also an issue of personal style.

File: gender_novels/novel.py (hunk 7 above)
Comment (ArrowAce) [reply gender_novels/novel.py]: I can put in the existing boiler plate remover but will comment it out and add a note, as there are certain books which actually do not have the standard project gutenberg boilerplate for which that function does not work (eg, james_highway.txt).

The gutenberg strip_headers() function does work on james_highway.txt however, so it is believed that strip_headers() works for all boilerplates, including nonstandard ones.

File: gender_novels/novel.py (hunk 9 above)
Comment (ArrowAce) [reply gender_novels/novel.py]: Thanks for the .split suggestion — goes to show how much easier things are with better knowledge of what's available.

I will look at the NLTK request and discuss that with them, as without line breaks there is no easy way to identify what lines were originally chapter titles.

File: gender_novels/novel.py (hunk 4 above)
Comment (ArrowAce) [reply gender_novels/novel.py]: What would the doctest look like for a function like this, which processes an entire novel?

Comment (mscuthbert) [reply gender_novels/novel.py]: Look at Doctest ellipsis for ways of making long tests shorter.  You can do:

>>> novel.text
It was a dark and stormy night...The End.
>>> len(novel.text)
892012

**Comment by mscuthbert** [comment]
@ArrowAce -- is the latest commit ready to review against Stephan's comments? Thanks!

**Comment by ArrowAce** [comment]
After a long discussion with Ife, Dina, and Stephan, we have concluded that removing the Table of Contents is not necessary: ToCs are very variable in format, making it difficult to code a way to remove all of them, and their inclusion will not affect things significantly. Moreover, on the analysis side of things, when the NLTK analysis breaks due to trying to parse ToC lines, we can try/catch those lines and discard those lines from our analysis (since they are non-English sentences).

The discussion about the ToC remover challenges, however, has been very useful and will be invaluable in the writeup.

In summary, I have thus decided to abandon the addition of a remove_table_of_contents function; however, the remove_boilerplate_text function updates are still an improvement that will be included.

**Comment by srisi** [comment]
Thanks!
